### PR TITLE
Fix some TCK problems

### PR DESCRIPTION
**Pull request by DarthMax**
This PR fixes several TCk failues:
- queries can start with OPTIONAL MATCH
- simplify logical planning of patterns
- fix planning of var length expand with empty intervals
- add support for `range`, `STARTS WITH`, `ENDS WITH`, `CONTAINS`

### CODE REVIEW

File: okapi-relational/src/main/scala/org/opencypher/okapi/relational/impl/graph/EmptyGraph.scala [hunk 1]
```diff
@@ -52,7 +54,17 @@ sealed case class EmptyGraph[T <: Table[T]](implicit val session: RelationalCyph
...
+    val variable = Var("")(entityType)
+
+    val exprs = entityType match {
+      case CTRelationship(types, _) =>
+        Seq(variable, StartNode(variable)(CTInteger), EndNode(variable)(CTInteger)) ++ types.map(t => HasType(variable, RelType(t))(CTBoolean))
+      case CTNode(labels, _) =>
+        Seq(variable) ++ labels.map(l => HasLabel(variable, Label(l))(CTBoolean))
+      case other => throw IllegalArgumentException("EntityType to be either CTNode or CTRelationship", other)
+    }
+    val scanHeader = RecordHeader.empty.withExprs(exprs.toSet)
```
Comment (s1ck) [inline okapi-relational/src/main/scala/org/opencypher/okapi/relational/impl/graph/EmptyGraph.scala]: Should empty graph maybe have an optional header argument instead? 
Another option would be a convenience method `RecordHeader.from(nodeType: CTNode)` and for rels

File: okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalPlanner.scala [hunk 2]
```diff
@@ -420,33 +420,13 @@ class LogicalPlanner(producer: LogicalOperatorProducer)
       val filteredPlan = planFilter(patternPlan, where)
       filteredPlan
     } else {
-      // TODO: Find a way to feed the same input into all arms of the cartesian product without recomputing it
-      val bases = plan +: components.map(_ => Start(plan.graph, SolvedQueryModel.empty)).tail
-      val plans = bases.zip(components).map {
+      components.foldLeft(plan) {
```
Comment (s1ck) [inline okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalPlanner.scala]: very nice simplification

File: okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalPlanner.scala [hunk 3]
```diff
@@ -498,7 +498,13 @@ class LogicalPlanner(producer: LogicalOperatorProducer)
           case _: DirectedVarLengthRelationship => Directed
           case _: UndirectedVarLengthRelationship => Undirected
         }
-        producer.planBoundedVarLengthExpand(c.source, r, c.target, v.edgeType, direction, v.lower, v.upper.get, sourcePlan, targetPlan)
+
+        if(v.upper.getOrElse(Integer.MAX_VALUE) < v.lower) {
```
Comment (s1ck) [inline okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalPlanner.scala]: missing space after `if`

File: okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala [hunk 4]
```diff
@@ -438,34 +446,38 @@ object SchemaTyper {
...
+    ): Eff[R, Set[FunctionSignature]]
   }
 
-  private case object FunctionInvocationTyper extends SignatureBasedInvocationTyper[FunctionInvocation] {
-    override protected def generateSignaturesFor[R: _hasSchema: _keepsErrors: _hasTracker: _logsTypes](
-        expr: FunctionInvocation,
-        args: Seq[(Expression, CypherType)]): Eff[R, Set[FunctionSignature]] =
-      expr.function match {
-
+  private case object BasicSignatureBasedTyper extends SignatureBasedInvocationTyper[Expression] {
```
Comment (s1ck) [inline okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala]: Why `Basic...` ?

File: okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala [hunk 5]
```diff
@@ -504,48 +516,53 @@ object SchemaTyper {
...
-
         case f: TypeSignatures =>
           val set = f.signatures.flatMap(_.convert).toSet
           pure(set)
-
-        case _ =>
-          wrong[R, TyperError](UnsupportedExpr(expr)) >> pure(Set.empty)
       }
+
+      case f: OperatorExpression =>
```
Comment (s1ck) [inline okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala]: nice generalization

Comment (DarthMax) [reply okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala]: It is only a fallback though, not all OperatorExpressions implement the signatures

File: okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala (hunk 4 above)
Comment (DarthMax) [reply okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala]: mhh there is an AddTyper as well, I could also call it `SignatureBasedInvocationTyperImpl`

File: okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalPlanner.scala (hunk 3 above)
Comment (DarthMax) [reply okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalPlanner.scala]: :+1:

File: okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala (hunk 5 above)
Comment (s1ck) [reply okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala]: too bad

File: okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala (hunk 4 above)
Comment (s1ck) [reply okapi-ir/src/main/scala/org/opencypher/okapi/ir/impl/typer/SchemaTyper.scala]: :+1:

File: okapi-relational/src/main/scala/org/opencypher/okapi/relational/impl/graph/EmptyGraph.scala (hunk 1 above)
Comment (DarthMax) [reply okapi-relational/src/main/scala/org/opencypher/okapi/relational/impl/graph/EmptyGraph.scala]: I went for the latter option

**Pull request merged**